File: _solutions/22_quiz-game-4/22_quiz-game-4.py
# %%
import chatlas
import dotenv
from pyhere import here

dotenv.load_dotenv()

# %%
from pathlib import Path
from typing import Any, Literal

from faicons import icon_svg
from playsound3 import playsound

SoundChoice = Literal["correct", "incorrect", "new-round", "you-win"]

sound_map: dict[SoundChoice, Path] = {
    "correct": here("data/sounds/smb_coin.wav"),
    "incorrect": here("data/sounds/wilhelm.wav"),
    "new-round": here("data/sounds/victory_fanfare_mono.wav"),
    "you-win": here("data/sounds/smb_stage_clear.wav"),
}

icon_map: dict[SoundChoice, Any] = {
    "correct": icon_svg("circle-check", fill="var(--bs-success)"),
    "incorrect": icon_svg("circle-xmark", fill="var(--bs-danger)"),
    "new-round": icon_svg("circle-play", fill="var(--bs-primary)"),
    "you-win": icon_svg("trophy", fill="var(--bs-warning)"),
}

title_map: dict[SoundChoice, str] = {
    "correct": "That's right!",
    "incorrect": "Oops, not quite.",
    "new-round": "Let's goooooooo!",
    "you-win": "You Win",
}


def play_sound(sound: SoundChoice = "correct") -> str:
    """
    Plays a sound effect.

    Parameters
    ----------
    sound: Which sound effect to play: "correct", "incorrect", "new-round" or
           "you-win". Play the "new-round" sound after the user picks a theme
           for the round. Play the "correct" and "incorrect" sounds when the
           user answers a question correctly or incorrectly, respectively. And
           play the "you-win" sound at the end of a round of questions.

    Returns
    -------
    A confirmation that the sound was played.
    """
    if sound not in sound_map.keys():
        raise ValueError(
            f"sound must be one of {sorted(sound_map.keys())}; got {sound!r}"
        )

    playsound(sound_map[sound])

    return chatlas.ContentToolResult(
        value=f"The '{sound}' sound was played.",
        extra={
            "display": {
                "title": title_map[sound],
                "icon": icon_map[sound],
            }
        },
    )


# %%
chat = chatlas.ChatAnthropic(
    model="claude-3-7-sonnet-20250219",
    system_prompt=here("_solutions/14_quiz-game-1/prompt.md").read_text(),
)

# %%
chat.register_tool(
    play_sound,
    annotations={"title": "Play Sound Effect"},
)

_ = chat.chat("Begin the quiz game.", echo="none")
chat.set_turns(chat.get_turns()[-1:])  # Keep only the assistant's greeting

# %%
chat.app()

File: _solutions/22_quiz-game-4/22_quiz-game-4.R
library(beepr)
library(ellmer)

#' Plays a sound effect.
#'
#' @param sound Which sound effect to play: `"correct"`, `"incorrect"`,
#'   `"new-round"`, or `"you-win"`.
#' @returns A confirmation that the sound was played.
play_sound <- function(
  sound = c("correct", "incorrect", "new-round", "you-win")
) {
  sound <- match.arg(sound)

  switch(
    sound,
    correct = beepr::beep("coin"),
    incorrect = beepr::beep("wilhelm"),
    "new-round" = beepr::beep("fanfare"),
    "you-win" = beepr::beep("complete")
  )

  icon <- switch(
    sound,
    correct = fontawesome::fa_i(
      "circle-check",
      class = "text-success",
      prefer_type = "solid"
    ),
    incorrect = fontawesome::fa_i(
      "circle-xmark",
      class = "text-danger",
      prefer_type = "solid"
    ),
    "new-round" = fontawesome::fa_i(
      "circle-play",
      class = "text-primary",
      prefer_type = "solid"
    ),
    "you-win" = fontawesome::fa_i("medal", class = "text-warning")
  )

  title <- switch(
    sound,
    correct = "That's right!",
    incorrect = "Oops, not quite.",
    "new-round" = "Let's goooooo!",
    "you-win" = "You Win!"
  )

  ContentToolResult(
    glue::glue("The '{sound}' sound was played."),
    extra = list(
      display = list(
        title = title,
        icon = icon
      )
    )
  )
}

tool_play_sound <- tool(
  play_sound,
  description = "Play a sound effect",
  arguments = list(
    sound = type_enum(
      c("correct", "incorrect", "new-round", "you-win"),
      description = paste(
        "Which sound effect to play.",
        "Play 'new-round' after the user picks a theme for the round.",
        "Play 'correct' or 'incorrect' after the user answers a question.",
        "Play 'you-win' at the end of a round of questions."
      )
    )
  ),
  annotations = tool_annotations(title = "Play Sound Effect")
)

chat <- chat(
  "anthropic/claude-3-7-sonnet-20250219",
  system_prompt = interpolate_file(
    # Replace `_solutions` with `_exercises` to get your own prompt from before
    here::here("_solutions/14_quiz-game-1/prompt.md")
  )
)

chat$register_tool(tool_play_sound)

. <- chat$chat("Begin", echo = FALSE) # get LLM to give us the greeting
chat$set_turns(chat$get_turns()[-1]) # remove first "Begin" message
live_browser(chat, quiet = TRUE)
